include/zephyr/bindesc.h

```diff
@@ -0,0 +1,398 @@
+ */
+#define DEFINE_BINDESC_STR(name, id, value)	\
+	DEFINE_BINDESC_ENTRY(name) = {		\
+		.tag = BINDESC_TAG(STR, id),	\
+		.len = (uint16_t)sizeof(value),	\
+		.data = value,			\
+	}
+
+/**
+ * @brief Define a binary descriptor of type uint.
+ *
+ * @details
+ * Define an integer that is registered in the binary descriptor header.
+ * The defined integer can be accessed using @ref BINDESC_GET_UINT
+ *
+ * @note The defined integer is not static, so its name must not collide with
+ * any other symbol in the executable.
+ *
+ * @param name Name of the descriptor
+ * @param id Unique ID of the descriptor
+ * @param value An integer value for the descriptor
+ */
+#define DEFINE_BINDESC_UINT(name, id, value)		\
+	DEFINE_BINDESC_ENTRY(name) = {			\
+		.tag = BINDESC_TAG(UINT, id),		\
+		.len = (uint16_t)sizeof(uint32_t),	\
+		.data = sys_uint32_to_array(value),	\
+	}
+
+/**
+ * @brief Define a binary descriptor of type bytes.
+ *
+ * @details
+ * Define a uint8_t array that is registered in the binary descriptor header.
+ * The defined array can be accessed using @ref BINDESC_GET_BYTES.
+ * The value should be given as an array literal, wrapped in parentheses, for
+ * example:
+ *
+ *     DEFINE_BINDESC_BYTES(name, id, ({1, 2, 3, 4}));
+ *
+ * @note The defined array is not static, so its name must not collide with
+ * any other symbol in the executable.
+ *
+ * @param name Name of the descriptor
+ * @param id Unique ID of the descriptor
+ * @param value A uint8_t array as data for the descriptor
+ */
+#define DEFINE_BINDESC_BYTES(name, id, value)				\
+	DEFINE_BINDESC_ENTRY(name) = {					\
+		.tag = BINDESC_TAG(BYTES, id),				\
+		.len = (uint16_t)sizeof((uint8_t [])__DEBRACKET value),	\
+		.data = __DEBRACKET value,				\
+	}
+
+/**
+ * @brief Get the value of a string binary descriptor
+ *
+ * @details
+ * Get the value of a string binary descriptor, previously defined by
+ * DEFINE_BINDESC_STR.
+ *
+ * @param name Name of the descriptor
+ */
+#define BINDESC_GET_STR(name) BINDESC_NAME(name).data
+
+/**
+ * @brief Get the value of a uint binary descriptor
+ *
+ * @details
+ * Get the value of a uint binary descriptor, previously defined by
+ * DEFINE_BINDESC_UINT.
+ *
+ * @param name Name of the descriptor
+ */
+#define BINDESC_GET_UINT(name) *(uint32_t *)&(BINDESC_NAME(name).data)
+
+/**
+ * @brief Get the value of a bytes binary descriptor
+ *
+ * @details
+ * Get the value of a string binary descriptor, previously defined by
+ * DEFINE_BINDESC_BYTES. The returned value can be accessed as an array:
+ *
+ *     for (size_t i = 0; i < BINDESC_GET_SIZE(name); i++)
+ *         BINDESC_GET_BYTES(name)[i];
+ *
+ * @param name Name of the descriptor
+ */
+#define BINDESC_GET_BYTES(name) BINDESC_NAME(name).data
+
+/**
+ * @brief Get the size of a binary descriptor
+ *
+ * @details
+ * Get the size of a binary descriptor. This is particularly useful for
+ * bytes binary descriptors where there's no null terminator.
```

Comment:
Why do you need null terminator in string, you know the length already because you have encoded it anyway.

Reply:
> Why do you need null terminator in string, you know the length already because you have encoded it anyway.

OK I see now, to access from the code.